components/service/glean/src/main/java/mozilla/components/service/glean/storages/GenericScalarStorageEngine.kt

```diff
@@ -0,0 +1,203 @@
+     *
+     * @return the [ScalarType] recorded in the requested store
+     */
+    @Synchronized
+    fun getSnapshot(storeName: String, clearStore: Boolean): GenericDataStorage<ScalarType>? {
+        val allLifetimes: GenericDataStorage<ScalarType> = mutableMapOf()
+
+        // Make sure data with "user" lifetime is loaded before getting the snapshot.
+        // We still need to catch exceptions here, as `getAll()` might throw.
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            userLifetimeStorage.all
+        } catch (e: NullPointerException) {
+            // Intentionally left blank. We just want to fall through.
+        }
+
+        // Get the metrics for all the supported lifetimes.
+        for (store in dataStores) {
+            store[storeName]?.let {
+                allLifetimes.putAll(it)
+            }
+        }
+
+        if (clearStore) {
+            // We only allow clearing metrics with the "ping" lifetime.
+            dataStores[Lifetime.Ping.ordinal].remove(storeName)
+        }
+
+        return if (allLifetimes.isNotEmpty()) allLifetimes else null
+    }
+
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clearStore the requested store
+     *
+     * @return the [JSONObject] containing the recorded data.
+     */
+    override fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any? {
+        return getSnapshot(storeName, clearStore)?.let { dataMap ->
+            return JSONObject(dataMap)
+        }
+    }
+
+    /**
+     * Helper function for the derived classes. It can be used to record
+     * simple scalars to the internal storage.
+     */
+    protected fun recordScalar(
+        stores: List<String>,
+        category: String,
+        name: String,
+        lifetime: Lifetime,
+        value: ScalarType
+    ) {
+        checkNotNull(applicationContext) { "No recording can take place without an application context" }
+
+        // Record a copy of the metric in all the needed stores.
+        @SuppressLint("CommitPrefEdits")
+        val userPrefs: SharedPreferences.Editor? =
+            if (lifetime == Lifetime.User) userLifetimeStorage.edit() else null
+        for (storeName in stores) {
+            val storeData = dataStores[lifetime.ordinal].getOrPut(storeName) { mutableMapOf() }
+            val entryName = "$category.$name"
+            storeData[entryName] = value
+            // Persist data with "user" lifetime
+            if (lifetime == Lifetime.User) {
+                userPrefs?.putString("$storeName#$entryName", serializeSingleMetric(value))
+            }
+        }
+        userPrefs?.apply()
+    }
+
+    @VisibleForTesting
+    internal open fun clearAllStores() {
+        for (store in dataStores) {
+            store.clear()
+        }
+    }
```

Comment:
nit: Some of the functions here you can "shrink" by using some of the various list extension functions, e.g.:

```Kotlin
internal open fun clearAllStores() = dataStores.forEach { it.clear() }
```